Upgrade Gradle to 7.4 & AGP to 7.1.1 (& Disable Jetifier)

ParaskP7 opened the pull request:
This PR is part of [Upgrade Gradle to 7.4 & AGP to 7.1.1 (& Disable Jetifier)](https://github.com/WordPress/gutenberg/pull/39097) PR. There are `13` libraries that are upgraded to `Gradle 7.4` & `AGP 7.1.1` (& `Disable Jetifier`).

-----

All those PRs follow the general outline below:
- `Gradle` version upgraded to `7.3.3` with the `./gradlew wrapper --gradle-version=7.3.3 --distribution-type=all` command.
- `AGP` version upgrade to `7.0.4` (see `android/settings.gradle.kts` change).
- Build output diff:
  - To identify and fix new warnings/errors that got added.
  - To identify and verify old warnings/errors that got removed.
- Lint output diff
  - To identify and fix new warnings/errors that got added.
  - To identify and verify old warnings/errors that got removed.
- The above 4 steps were redone with a follow-up update to `Gradle 7.4` & `AGP 7.1.1`.
- Jetifier drop (see `android/gradle.properties` change).
- **OUTDATED** ❌
  - Version bump to `1.9.0-wp-3` (see `package.json`).
  - Tarball update to `1.9.0-wp-3` (see `react-native-reanimated-1.9.0-wp-3.tgz`).
  - The build was also verified through [Jitpack](https://jitpack.io/#wordpress-mobile/react-native-reanimated) (see [build.log](https://jitpack.io/com/github/wordpress-mobile/react-native-reanimated/update~gradle-to-7.3.3-agp-to-7.0.4-1.9.0-gec87aa0-29/build.log)).
- **OUTDATED** ❌
  - `AGP` version upgrade to `7.1.1` (from `4.2.2`) (within `android/build.gradle`). Cc @oguzkocer
  - Version bump to `2.2.4-wp-2` (see `package.json`).
  - Tarball update to `2.2.4-wp-2` (see `react-native-reanimated-2.2.4-wp-2.tgz`).
  - The build was also verified through [Jitpack](https://jitpack.io/#wordpress-mobile/react-native-reanimated) (see [build.log](https://jitpack.io/com/github/wordpress-mobile/react-native-reanimated/0d5474cb8e/build.log)).
- **UPDATED** ✅
  - `Spotless` configuration change (within `settings.gradle.kts` and `android/gradle`).
  - Version bump to `2.4.1-wp-2` (see `package.json`).
  - Tarball update to `2.4.1-wp-2` (see `react-native-reanimated-2.4.1-wp-2.tgz`).
  - The build was also verified through [Jitpack](https://jitpack.io/#wordpress-mobile/react-native-reanimated) (see [build.log](https://jitpack.io/com/github/wordpress-mobile/react-native-reanimated/734d446671/build.log)).

-----

**To test - Now**

These changes can be tested as part of the [gutenberg PR](https://github.com/WordPress/gutenberg/pull/39097) which is [updated to use the temporarily non-tagged node module project dependencies](https://github.com/WordPress/gutenberg/pull/39097/commits/d10243401c1ef73be0bcc74023adba25c4787b56) generated with these changes.

**To test - Later** ⚙️

These changes can be tested as part of the [gutenberg-mobile PR ⚙️](TODO) or [WordPress-Android PR ⚙️](TODO) which is [updated to use the bundle ⚙️](TODO) generated with these changes.

FYI: It's best to leave the testing step to the [gutenberg PR](https://github.com/WordPress/gutenberg/pull/39097) review since even if these PRs are merged, they won't impact anything until [the package.json file is updated](https://github.com/WordPress/gutenberg/pull/39097/commits/d10243401c1ef73be0bcc74023adba25c4787b56). Worst case scenario, a follow up PR will get opened to fix any issues that are found during testing.

-----

**Follow up**

Once these PRs are merged in, a new tag will get created for each library and [the package.json file will be again updated accordingly](https://github.com/WordPress/gutenberg/pull/39097/commits/d10243401c1ef73be0bcc74023adba25c4787b56).

ParaskP7 committed:
Build: Update gradle to 7.3.3

ParaskP7 committed:
Build: Update agp to 7.0.4

ParaskP7 commented:
👋 @oguzkocer !

Do you have any idea why the `Autosquash` checks fail, what are they, or why are they utilized? 🤔 

I get the below and I am not sure what I need to be doing with the `github_token`, which token, etc:
```
Run tibdex/autosquash@v2
Error: Error: Input required and not supplied: github_token
```

Many thanks for your support and guidance! 🙏

oguzkocer commented:
> I get the below and I am not sure what I need to be doing with the github_token, which token, etc:

[The workflow seems to be from the original repo](https://github.com/wordpress-mobile/react-native-reanimated/blob/wp-fork/.github/workflows/autosquash.yml) and [the secret](https://github.com/wordpress-mobile/react-native-reanimated/blob/wp-fork/.github/workflows/autosquash.yml#L39) is not configured in this repo. I think it's safe to ignore it.

ParaskP7 commented:
> [The workflow seems to be from the original repo](https://github.com/wordpress-mobile/react-native-reanimated/blob/wp-fork/.github/workflows/autosquash.yml) and [the secret](https://github.com/wordpress-mobile/react-native-reanimated/blob/wp-fork/.github/workflows/autosquash.yml#L39) is not configured in this repo. I think it's safe to ignore it.

Ignore it it is then, at least for now, thanks Oguz! 🙇

ParaskP7 committed:
Build: Update gradle to 7.4

ParaskP7 committed:
Build: Update agp to 7.1.1

ParaskP7 committed:
Build: Disable jetifier

The 'com.dipien.byebyejetifier' plugin was used to determine whether
Jetifier can be disable for this project.

After configuring 'Bye Bye Jetifier' and running the below command:
./gradlew canISayByeByeJetifier -Pandroid.enableJetifier=false

The output was clear, Jetifier can be now safely disabled for this
project: > Task :canISayByeByeJetifier

=========================================
Project: :
=========================================
 * No legacy android support usages found

===================================================== ...
* No dependencies with legacy android support usages! ...
===================================================== ...

... ===============================
... You can say Bye Bye Jetifier. *
... ===============================

ParaskP7 committed:
Build: Bump version to 1.9.0-wp-3

ParaskP7 committed:
Build: Generate new tarball for 1.9.0-wp-3

fluiddot committed:
Merge branch 'wp-fork' into update/gradle-to-7.3.3-agp-to-7.0.4

# Conflicts:
#	package.json

fluiddot committed:
Generate tarball

ParaskP7 committed:
Build: Update agp to 7.1.1

fluiddot committed:
Use gradlew in createNPMpackage

fluiddot committed:
Run externalNativeBuildCleanDebug task in create NPM package script

fluiddot committed:
Generate new AAR

fluiddot committed:
Generate tarball

ParaskP7 committed:
Build: Move agp and kotlin dependencies to plugin management

ParaskP7 committed:
Build: Re-add maven-publish plugin configuration

ParaskP7 committed:
Build: Generate new aar for 66-hermes

ParaskP7 committed:
Build: Generate new tarball for 2.2.4-wp-2

ParaskP7 commented:
👋 @oguzkocer !

As done [here](https://github.com/wordpress-mobile/react-native-gesture-handler/pull/12#issuecomment-1062858343), Can you please take a look at this b4be54efad2c19da875ce02ea2211861ac3156d8 commit, which resolves the same issue, and verify that I am not missing anything? 🙇

Cc @fluiddot

PS: Can you also verify this 9bd99f5e5f54b37dbc793f39223f88fd9d29c5cb commit, were I am re-adding the maven-publish plugin configuration. 🤔

fluiddot commented on android/build.gradle:
```diff
@@ -478,4 +466,16 @@ tasks.whenTaskAdded { task ->
\ No newline at end of file
+}
+
+afterEvaluate {
+    publishing {
+        publications {
+            release(MavenPublication) {
+                from components.release
+                groupId = 'com.github.wordpress-mobile'
+                artifactId = 'react-native-reanimated'
+            }
+        }
+    }
+}
```
The Android project we're publishing is located in `android-npm` ([reference](https://github.com/wordpress-mobile/react-native-reanimated/blob/ae08ebd3ad6ac863dbde916bf80050ba26a66506/android-npm/build.gradle#L21-L32)), this one is only used for generating the AAR (`android-npm/react-native-reanimated-66-hermes.aar`) and is invoked when running `./createNPMPackage.sh`. So, in theory, we don't need to add part.

ParaskP7 replied:
Oh cool, I am going to revert that change then, thanks for letting me know @fluiddot ! 🙇

fluiddot replied:
@ParaskP7 Thanks for addressing this 🙇 .

ParaskP7 committed:
Build: Remove maven-publish plugin configuration

The Android project publishing is now located in 'android-npm', it is
only used for generating the AAR and is invoked when running the
'./createNPMPackage.sh' script.

oguzkocer commented:
@ParaskP7 b4be54efad2c19da875ce02ea2211861ac3156d8 looks good to me.

fluiddot commented:
@ParaskP7 Heads up that I'm planning to update the version of the package to `2.4.1` in https://github.com/wordpress-mobile/react-native-reanimated/pull/13. Once I proceed with the merge, I'll update this PR and the companion Gutenberg PR.

ParaskP7 commented:
👋 @fluiddot !

> @ParaskP7 Heads up that I'm planning to update the version of the package to 2.4.1 in https://github.com/wordpress-mobile/react-native-reanimated/pull/13. Once I proceed with the merge, I'll update this PR and the companion Gutenberg PR.

Thanks for the heads-up and for handling the update on this PR later-on as well, much appreciated! 🙇

fluiddot committed:
Merge branch 'wp-fork' into update/gradle-to-7.3.3-agp-to-7.0.4

# Conflicts:
#	android-npm/react-native-reanimated-66-hermes.aar
#	android/build.gradle
#	createNPMPackage.sh
#	package.json

fluiddot commented on android/build.gradle:
```diff
@@ -1,15 +1,26 @@
+    id 'com.android.library'
+    id 'maven-publish'
+    id 'de.undercouch.download'
+    id 'com.diffplug.gradle.spotless' apply false
+}
+
+if (project == rootProject) {
+    apply from: 'spotless.gradle'
+}
```
@ParaskP7 this plugin has been added in version `2.4.1` ([reference](https://github.com/software-mansion/react-native-reanimated/blob/e25d29aa689a1442d298927303473da44a07d475/android/build.gradle#L372-L374)), but it's only meant to be applied when building the library. Let me know if this is the proper way to apply it, thanks 🙇.

ParaskP7 replied:
👋 @fluiddot !

I am not sure why this `Spotless` plugin needs to be guarded with this `if (project == rootProject) { ... }` check, but we could maybe remove that check altogether and just have `Spotless` enabled by default. Then, this configuration will become much more straightforward. Wdyt? 🤔 

Do do so, we would need to:
1. Use `com.diffplug.spotless` instead of `com.diffplug.gradle.spotless` when defying the plugin version (`settings.gradle.kts`).
2. Use `id 'com.diffplug.spotless'` instead of `id 'com.diffplug.gradle.spotless' apply false` when applying the plugin (`build.gradle`).
3. Since the `spotless.gradle` is only used within the above `build.gradle` file, to move the `spotless { ... }` configuration directly into the `build.gradle` above and removing the `spotless.gradle` file altogether.

fluiddot replied:
> I am not sure why this Spotless plugin needs to be guarded with this if (project == rootProject) { ... } check, but we could maybe remove that check altogether and just have Spotless enabled by default. Then, this configuration will become much more straightforward. Wdyt? 🤔

Me neither to be honest, but I agree about enabling it by default. I'll proceed with the changes, thanks 🙇 .

fluiddot replied:
@ParaskP7 I've applied your suggestion in https://github.com/wordpress-mobile/react-native-reanimated/pull/9/commits/734d44667145464f4006294b94644e929b53f28c.

ParaskP7 replied:
Awesome @fluiddot , I double-checked that and can verify this change, thank you!

PS: Now we need to update the `aar` and `tarball` once more and we are done, right? 🤔

fluiddot replied:
> PS: Now we need to update the aar and tarball once more and we are done, right? 🤔

Yep, actually, I ran the `createNPMPackage` script in https://github.com/wordpress-mobile/react-native-reanimated/commit/734d44667145464f4006294b94644e929b53f28c to generate both AAR and tarball files but there were no modifications.

fluiddot replied:
Btw, I also updated the reference of this package in the Gutenberg PR ([reference](https://github.com/WordPress/gutenberg/pull/39097/commits/4e9762762ba54f879beaebb8bba3007b5eac75cb)).

ParaskP7 replied:
> Yep, actually, I ran the createNPMPackage script in https://github.com/wordpress-mobile/react-native-reanimated/commit/734d44667145464f4006294b94644e929b53f28c to generate both AAR and tarball files but there were no modifications.

Nice! 👍 

> Btw, I also updated the [reference](https://github.com/WordPress/gutenberg/pull/39097/commits/4e9762762ba54f879beaebb8bba3007b5eac75cb) of this package in the Gutenberg PR (reference).

Awesome! 💯

fluiddot committed:
Merge branch 'wp-fork' into update/gradle-to-7.3.3-agp-to-7.0.4

fluiddot committed:
Fox spotless plugin integration

ParaskP7 commented:
Since we removed the source code for these projects from `react-native-bridge`, we don’t need these updates to happen together with the other projects anymore. That means we can close this draft PR for now and reopen it, if need be, in the future. There is little value in merging this now and diverging from the source repo.

PS: Eventually, we might upgrade `Gradle & AGP` when using a newer version of `React Native` if the new RN version requires it.
